From a7f87ae9f63aa1a3b5d0b54525b32e3b5a2dd043 Mon Sep 17 00:00:00 2001
From: Sandy Huang <hjc@rock-chips.com>
Date: Tue, 18 Sep 2018 14:51:47 +0800
Subject: [PATCH] drm/rockchip: vop: fix kernel panic when calc bandwidth

error log:

[   35.638112] Internal error: Accessing user space memory outside uaccess.h routines: 96000005 [#1] PREEMPT SMP
[   35.638996] Modules linked in:
[   35.639287] CPU: 3 PID: 226 Comm: surfaceflinger Not tainted 4.4.126
[   35.639873] Hardware name: Strides RK3399 Development Q7 Board Rev
[   35.640483] task: ffffffc0f100ec00 task.stack:ffffffc0eeaac000
[   35.641011] PC is at vop_crtc_bandwidth+0x22c/0x310
[   35.641448] LR is at vop_crtc_bandwidth+0x118/0x310
[   35.641883] pc : [<ffffff80084af988>] lr :[<ffffff80084af874>] pstate: 60400145
[   35.642526] sp : ffffffc0eeaafad0
[   35.642821] x29: ffffffc0eeaafad0 x28: 0000000000000020
...
[   35.780538] [<ffffff80084af988>] vop_crtc_bandwidth+0x22c/0x310
[   35.781057] [<ffffff80084a87b4>] rockchip_drm_atomic_commit+0xb0/0x1d0
[   35.781634] [<ffffff8008491678>] drm_atomic_commit+0x64/0x70
[   35.782140] [<ffffff8008492b08>] drm_mode_atomic_ioctl+0x5d0/0x6a0
[   35.782683] [<ffffff8008476b30>] drm_ioctl+0x2e4/0x400
[   35.783145] [<ffffff80081d0db0>] do_vfs_ioctl+0xa4/0x7d8
[   35.783616] [<ffffff80081d1540>] SyS_ioctl+0x5c/0x8c
[   35.784055] [<ffffff80080832f0>] el0_svc_naked+0x24/0x28
[   35.784528] Code: 7100113f 54ffff41 f9412ca0 f9400000 (f9400002)
[   35.785256] ---[ end trace b3812e3405f44f95 ]---
...

Change-Id: I695c9628b155294adcf5135f82e4af901ce1b82c
Signed-off-by: Sandy Huang <hjc@rock-chips.com>
---
 drivers/gpu/drm/rockchip/rockchip_drm_vop.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/drivers/gpu/drm/rockchip/rockchip_drm_vop.c b/drivers/gpu/drm/rockchip/rockchip_drm_vop.c
index aa46167a3cbd..f3613021b1f0 100644
--- a/drivers/gpu/drm/rockchip/rockchip_drm_vop.c
+++ b/drivers/gpu/drm/rockchip/rockchip_drm_vop.c
@@ -2345,7 +2345,8 @@ static size_t vop_plane_line_bandwidth(struct drm_plane_state *pstate)
 	int vskiplines = scl_get_vskiplines(src_height, dest_height);
 	size_t bandwidth;
 
-	if (!src_width || !src_height || !dest_width || !dest_height)
+	if (src_width <= 0 || src_height <= 0 || dest_width <= 0 ||
+	    dest_height <= 0)
 		return 0;
 
 	bandwidth = src_width * bpp / 8;
-- 
2.35.3

